src/internal/clusterstate/v2.10.0/pipelines_test.go

```diff
@@ -0,0 +1,175 @@
+package v2_10_0_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/clusterstate"
+	v2_10_0 "github.com/pachyderm/pachyderm/v2/src/internal/clusterstate/v2.10.0"
+	"github.com/pachyderm/pachyderm/v2/src/internal/migrations"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachd"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+	"github.com/pachyderm/pachyderm/v2/src/pps"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+func TestPipelineVersionsDeduplication(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping benchmark in short mode")
+	}
+	// - setup database with collections.pipelines, collections.jobs
+	// - migrate pipeline A from versions [1, 1, 2] -> [1, 2, 3]
+	type version int
+	type testCase struct {
+		desc          string
+		initial       []version
+		hasDuplicates bool
+	}
+	var (
+		tcs = []testCase{
+			{
+				desc:          "healthy pipeline versions",
+				initial:       []version{1, 2},
+				hasDuplicates: false,
+			},
+			{
+				desc:          "pipeline version duplicates",
+				initial:       []version{1, 1},
+				hasDuplicates: true,
+			},
+			{
+				desc:          "pipeline version duplication at beginning",
+				initial:       []version{1, 1, 2},
+				hasDuplicates: true,
+			},
+			{
+				desc:          "pipeline version duplication in the middle",
+				initial:       []version{1, 2, 2, 3},
+				hasDuplicates: true,
+			},
+			{
+				desc:          "pipeline version duplication at the end",
+				initial:       []version{1, 2, 3, 3},
+				hasDuplicates: true,
+			},
```

Comment:
I'm not sure the algorithm works if there are TWO duplicate versions, but we have to assume that such a case might be lurking.  We should test that.

Reply:
added a test